fbw-common/src/wasm/systems/systems/src/surveillance/egpws/runtime.rs

```diff
@@ -0,0 +1,1286 @@
+                        .powf(-context.delta_as_secs_f64() / Self::MODE_4_FILTER_TIME_CONSTANT)))
+            .max(0.);
+        } else {
+            self.mode_4_c_filter_value = 0.;
+        }
+
+        let mode_4_c_boundary = self
+            .mode_4_c_filter_value
+            .clamp(30., self.mode_4_a_alert_upper_boundary);
+
+        let mode_4_c_enabled = !self.on_ground
+            && self.flight_phase == FlightPhase::Takeoff
+            && !(discrete_inputs.landing_flaps && discrete_inputs.landing_gear_downlocked)
+            && self.ra_ft < 1000.;
+
+        let mode_4_c_lamp_active =
+            mode_4_c_enabled && mode_4_c_boundary >= self.ra_ft && self.ra_ft > 100.;
+
+        // If the audio declutter is disabled, the aural warning is always active when in the alert area.
+        // Otherwise, increase the declutter threshold (ratcheting) once the aural warning has played once.
+        if self.pin_programs.audio_declutter_disable || !self.mode_4_lamp_active {
+            self.mode_4_c_declutter_threshold_increase = 0.;
+        } else if self.mode_4_too_low_terrain_voice_active
+            && self.aural_output == AuralWarning::TooLowTerrain
+            && self.number_of_aural_warning_emissions > 0
+        {
+            self.mode_4_c_declutter_threshold_increase += 0.2;
+        }
+
+        let biased_ra_ft = self.ra_ft * (1. + self.mode_4_c_declutter_threshold_increase);
+        let mode_4_c_too_low_terrain_voice_active =
+            mode_4_c_enabled && mode_4_c_boundary >= biased_ra_ft && biased_ra_ft > 100.;
+
+        //println!("Mode 4: ra: {:.1} ft, filter: {:.1} ft, upper boundary: {:.1} ft, biased ra: {:.1} ft, enabled: {},
+        // lamp active: {}, too low terrain voice active: {}",
+        //         self.ra_ft,
+        //         self.mode_4_c_filter_value,
+        //         self.mode_4_a_alert_upper_boundary,
+        //         biased_ra_ft,
+        //         mode_4_c_enabled,
+        //         mode_4_c_lamp_active,
+        //         mode_4_c_too_low_terrain_voice_active,
+        //     );
+
+        // Just write to the output, as mode 4C cannot be active at the same time as 4AB
+        self.mode_4_lamp_active = mode_4_c_lamp_active || mode_4_ab_lamp_active;
+        self.mode_4_too_low_terrain_voice_active =
+            mode_4_c_too_low_terrain_voice_active || mode_4_ab_too_low_terrain_voice_active;
+    }
+
+    fn update_mode_5_logic(
+        &mut self,
+        context: &UpdateContext,
+        ils: &impl InstrumentLandingSystemBus,
+        ir: &impl InertialReferenceBus,
+        discrete_inputs: &TerrainAwarenessWarningSystemDiscreteInputs,
+    ) {
+        // Not sure if this is only when a G/S warning is active, or always
+        if discrete_inputs.gs_cancel && self.ra_ft < 2000. {
+            self.mode_5_cancel_active = true;
+        } else if self.ra_ft >= 2000. || self.ra_ft < 30. {
+            self.mode_5_cancel_active = false;
+        }
+
+        // 0.0875 is 1 dot
```

Comment:
This comment is incomplete since it is only for the gs deviation but not the loc deviation. Both of them are used after this comment though.